src/lib/type_comparison.hpp

```diff
@@ -122,16 +122,20 @@ void with_comparator_light(const PredicateCondition predicate_condition, const F
 template <typename Functor, typename DataType>
 void with_between_comparator(const PredicateCondition predicate_condition, const DataType& lower_value,
                              const DataType& upper_value, const Functor& func) {
+  // For integral types, we can assume that lower <= upper and therefore save one comparison by rearranging
+  // the equation: (x >= a && x < b) === ((x - a) < (b - a)); cf. https://stackoverflow.com/a/17095534/2204581
+  // This is quite a bit faster.
   if constexpr (std::is_integral_v<DataType>) {
     const auto upd_lower_value = is_lower_inclusive_between(predicate_condition) ? lower_value : lower_value + 1;
     const auto upd_upper_value = is_upper_inclusive_between(predicate_condition) ? upper_value : upper_value - 1;
 
-    Assert(lower_value <= upper_value, "Query asks for between predicate with empty range. This should have been filtered out earlier.");
+    DebugAssert(upd_lower_value <= upd_upper_value,
+                "Query asks for between predicate with empty range. This should have been filtered out earlier.");
 
     using UnsignedDataType = std::make_unsigned_t<DataType>;
-    const UnsignedDataType value_difference = upper_value - lower_value;
+    const UnsignedDataType value_difference = upd_upper_value - upd_lower_value;
 
-    return func([&upd_lower_value, &upd_upper_value](const DataType& value) {
+    return func([&upd_lower_value, &value_difference](const DataType& value) {
```

Comment:
We are in the case where `DataType` is an integral (i.e., `int32_t` or `int64_t`), and `UnsignedDataType` is its unsigned counterpart. References don't really make sense here. The compiler should take care for this (and hopefully inline anyway), but let's just copy here.